beacon-chain/blockchain/process_block.go

```diff
@@ -218,45 +218,50 @@ func (s *Service) onBlock(ctx context.Context, signed interfaces.SignedBeaconBlo
 
 	func() {
 		// Send notification of light client optimistic update to the state feed.
-		// Get the attestedHeader
-		prevSlot := slots.PrevSlot(postState.Slot())
-		prevBlockRoot := bytesutil.ToBytes32(postState.BlockRoots()[prevSlot%params.BeaconConfig().SlotsPerHistoricalRoot])
-		prevBlock, err := s.cfg.BeaconDB.Block(ctx, prevBlockRoot)
+		// Determine slots per period
+		config := params.BeaconConfig()
+		slotsPerPeriod := uint64(config.EpochsPerSyncCommitteePeriod) * uint64(config.SlotsPerEpoch)
+
+		// Get attested state
+		attestedRoot := signed.Block().ParentRoot()
+		attestedState, err := s.cfg.StateGen.StateByRoot(ctx, attestedRoot)
 		if err != nil {
-			log.WithError(err).Error("Could not get previous block")
+			log.WithError(err).Error("Could not get attested state")
 			return
 		}
 
-		prevBlockSignedHeader, err := prevBlock.Header()
-		if err != nil {
-			log.WithError(err).Error("Could not get previous block header")
-			return
+		// Get finalized block
+		var finalizedBlock interfaces.SignedBeaconBlock
+		finalizedCheckPoint := attestedState.FinalizedCheckpoint()
+		if finalizedCheckPoint != nil {
+			finalizedRoot := bytesutil.ToBytes32(finalizedCheckPoint.Root)
+			finalizedBlock, err = s.cfg.BeaconDB.Block(ctx, finalizedRoot)
+			if err != nil {
+				finalizedBlock = nil
+			}
 		}
 
-		attestedHeader := migration.V1Alpha1SignedHeaderToV1(prevBlockSignedHeader).GetMessage()
+		update, err := lightclienthelpers.CreateLightClientUpdate(
```

Comment:
We add some helper functions which implement functions based on spec pseudo codes.  

These helper functions are put in a different pkg, can be shared in both `endpoint` and `event stream` path.